pkg/otlp/model/translator/statspayload.go

```diff
@@ -0,0 +1,471 @@
+// also not be left empty in case of failure to obtain it, because empty has special meaning. An empty hostname means
+// that we are in a Lambda environment. Thus, we must use a placeholder.
+const UnsetHostnamePlaceholder = "__unset__"
+
+// StatsPayloadToMetrics converts OTLP Metrics to an APM Stats Payload.
+func (t *Translator) StatsPayloadFromMetrics(mx pmetric.Metrics) (pb.StatsPayload, error) {
+	var stats pb.StatsPayload
+	rmxs := mx.ResourceMetrics()
+	for i := 0; i < rmxs.Len(); i++ {
+		rmx := rmxs.At(i)
+		attr := rmx.Resource().Attributes()
+		if v, ok := attr.Get(keyAPMStats); !ok || !v.Bool() {
+			t.logger.Debug(fmt.Sprintf("StatsPayload Translator: was asked to convert metrics to stats payload, but identifier key %q was not present. Skipping.", keyAPMStats))
+			continue
+		}
+		hostname := getStr(attr, statsKeyHostname)
+		tags := strings.Split(getStr(attr, statsKeyTags), ",")
+		if hostname == UnsetHostnamePlaceholder {
+			src, err := t.source(attr)
+			if err != nil {
+				return pb.StatsPayload{}, err
+			}
+			switch src.Kind {
+			case source.HostnameKind:
+				hostname = src.Identifier
+			case source.AWSECSFargateKind:
+				hostname = ""
+				tags = append(tags, src.Tag())
+			}
+		}
+		cp := pb.ClientStatsPayload{
+			Hostname:         hostname,
+			Env:              getStr(attr, statsKeyEnv),
+			Version:          getStr(attr, statsKeyVersion),
+			Lang:             getStr(attr, statsKeyLang),
+			TracerVersion:    getStr(attr, statsKeyTracerVersion),
+			RuntimeID:        getStr(attr, statsKeyRuntimeID),
+			Sequence:         getInt(attr, statsKeySequence),
+			AgentAggregation: getStr(attr, statsKeyAgentAggregation),
+			Service:          getStr(attr, statsKeyService),
+			ContainerID:      getStr(attr, statsKeyContainerID),
+			Tags:             tags,
+		}
+		smxs := rmx.ScopeMetrics()
+		for j := 0; j < smxs.Len(); j++ {
+			mxs := smxs.At(j).Metrics()
+			var (
+				buck pb.ClientStatsBucket
+				agg  aggregations
+			)
+			for k := 0; k < mxs.Len(); k++ {
+				m := mxs.At(k)
+				switch m.Type() {
+				case pmetric.MetricTypeSum:
+					key, val := t.extractSum(m.Sum(), &buck)
+					switch m.Name() {
+					case "hits":
+						agg.Value(key).Hits = val
+					case "errors":
+						agg.Value(key).Errors = val
+					case "duration":
+						agg.Value(key).Duration = val
+					case "top_level_hits":
+						agg.Value(key).TopLevelHits = val
+					}
+				case pmetric.MetricTypeExponentialHistogram:
+					key, val := t.extractSketch(m.ExponentialHistogram(), &buck)
+					switch m.Name() {
+					case "ok_summary":
+						agg.Value(key).OkSummary = val
+					case "error_summary":
+						agg.Value(key).ErrorSummary = val
+					}
+				default:
+					t.logger.Debug(fmt.Sprintf(`StatsPayload Translator: metric named %q in Stats Payload should be of type "Sum" or "ExponentialHistogram" but is %q instead`, m.Name(), m.Type()))
```

Comment:
If any of these assumptions do not hold, something has gone really wrong. I think it would make our life easier if we fail more loudly (say, by returning the error and dropping the whole payload), since we will be able to detect these errors more easily in the wild